File: packages/runar-zig/src/sdk_unsound_primitives_test.zig
//! R-062 / CL-BUG-105 — see sdk_errors.zig for the finding.
//!
//! Mirrors packages/runar-sdk/src/__tests__/unsound-primitives.test.ts.

const std = @import("std");
const errors_mod = @import("sdk_errors.zig");
const types = @import("sdk_types.zig");

test "R-062: an ordinary artifact deploys with or without an acknowledgement" {
    const none: []const []const u8 = &.{};
    const some: []const []const u8 = &.{"verifySP1FRI"};
    try errors_mod.assertUnsoundPrimitivesAcknowledged(&.{}, none, "Counter.deploy");
    try errors_mod.assertUnsoundPrimitivesAcknowledged(&.{}, some, "Counter.deploy");
}

test "R-062: an unsound artifact is refused without an acknowledgement" {
    errors_mod.last_unsound = null;
    const declared: []const []const u8 = &.{"verifySP1FRI"};
    const none: []const []const u8 = &.{};
    try std.testing.expectError(
        error.UnsoundPrimitiveNotAcknowledged,
        errors_mod.assertUnsoundPrimitivesAcknowledged(declared, none, "Sp1Rollup.deploy"),
    );
    const rec = errors_mod.last_unsound orelse return error.TestExpectedRecordedContext;
    try std.testing.expectEqualStrings("verifySP1FRI", rec.primitiveSlice());
    try std.testing.expectEqualStrings("Sp1Rollup.deploy", rec.contextSlice());
}

test "R-062: the acknowledgement must name every primitive" {
    const one: []const []const u8 = &.{"verifySP1FRI"};
    try errors_mod.assertUnsoundPrimitivesAcknowledged(one, one, "Sp1Rollup.deploy");

    const two: []const []const u8 = &.{ "verifySP1FRI", "someFutureStub" };
    try std.testing.expectError(
        error.UnsoundPrimitiveNotAcknowledged,
        errors_mod.assertUnsoundPrimitivesAcknowledged(two, one, "Sp1Rollup.deploy"),
    );
    const rec = errors_mod.last_unsound orelse return error.TestExpectedRecordedContext;
    try std.testing.expectEqualStrings("someFutureStub", rec.primitiveSlice());

    const other: []const []const u8 = &.{"somethingElse"};
    try std.testing.expectError(
        error.UnsoundPrimitiveNotAcknowledged,
        errors_mod.assertUnsoundPrimitivesAcknowledged(one, other, "Sp1Rollup.deploy"),
    );
}

test "R-062: the marker survives artifact JSON parsing, and is absent otherwise" {
    const a = std.testing.allocator;

    var marked = try types.RunarArtifact.fromJson(a,
        \\{"version":"v","contractName":"Sp1Rollup","script":"51","unsoundPrimitives":["verifySP1FRI"]}
    );
    defer marked.deinit();
    try std.testing.expectEqual(@as(usize, 1), marked.unsound_primitives.len);
    try std.testing.expectEqualStrings("verifySP1FRI", marked.unsound_primitives[0]);

    var plain = try types.RunarArtifact.fromJson(a,
        \\{"version":"v","contractName":"Counter","script":"51"}
    );
    defer plain.deinit();
    try std.testing.expectEqual(@as(usize, 0), plain.unsound_primitives.len);
}

File: packages/runar-go/sdk_unsound_primitives_test.go
package runar

import (
	"strings"
	"testing"
)

// R-062 / CL-BUG-105 — see sdk_unsound_primitives.go for the finding.
// Mirrors packages/runar-sdk/src/__tests__/unsound-primitives.test.ts.

func unsoundArtifact(primitives ...string) *RunarArtifact {
	return &RunarArtifact{
		Version:           "runar-v1.0.0-rc.1",
		CompilerVersion:   "1.0.0-rc.1-go",
		ContractName:      "Sp1Rollup",
		Script:            "51",
		ASM:               "OP_1",
		BuildTimestamp:    "2026-09-13T00:00:00Z",
		UnsoundPrimitives: primitives,
	}
}

func TestUnsoundGuard_OrdinaryArtifactPassesEitherWay(t *testing.T) {
	plain := unsoundArtifact()
	for _, ack := range [][]string{nil, {}, {"verifySP1FRI"}} {
		if err := assertUnsoundPrimitivesAcknowledged(plain, ack, "Counter.Deploy"); err != nil {
			t.Fatalf("ordinary artifact must deploy freely; got %v", err)
		}
	}
	if err := assertUnsoundPrimitivesAcknowledged(nil, nil, "Counter.Deploy"); err != nil {
		t.Fatalf("a nil artifact must not panic or refuse here; got %v", err)
	}
}

func TestUnsoundGuard_RefusesUnacknowledged(t *testing.T) {
	err := assertUnsoundPrimitivesAcknowledged(unsoundArtifact("verifySP1FRI"), nil, "Sp1Rollup.Deploy")
	if err == nil {
		t.Fatal("an unsound artifact must NOT deploy without acknowledgement")
	}
	if !strings.Contains(err.Error(), "verifySP1FRI") {
		t.Errorf("error must name the primitive; got %v", err)
	}
	if !strings.Contains(err.Error(), "Sp1Rollup.Deploy") {
		t.Errorf("error must name the call site; got %v", err)
	}
	if !strings.Contains(err.Error(), "AcknowledgeUnsound") {
		t.Errorf("error must say how to proceed; got %v", err)
	}
}

func TestUnsoundGuard_AcknowledgementMustNameEveryPrimitive(t *testing.T) {
	if err := assertUnsoundPrimitivesAcknowledged(
		unsoundArtifact("verifySP1FRI"), []string{"verifySP1FRI"}, "Sp1Rollup.Deploy",
	); err != nil {
		t.Fatalf("a complete acknowledgement must be accepted; got %v", err)
	}

	err := assertUnsoundPrimitivesAcknowledged(
		unsoundArtifact("verifySP1FRI", "someFutureStub"), []string{"verifySP1FRI"}, "Sp1Rollup.Deploy",
	)
	if err == nil {
		t.Fatal("a PARTIAL acknowledgement must be refused")
	}
	if !strings.Contains(err.Error(), "someFutureStub") {
		t.Errorf("error must name the unacknowledged one; got %v", err)
	}

	if err := assertUnsoundPrimitivesAcknowledged(
		unsoundArtifact("verifySP1FRI"), []string{"somethingElse"}, "Sp1Rollup.Deploy",
	); err == nil {
		t.Fatal("acknowledging a DIFFERENT primitive must not unlock this one")
	}
}
